### DISCUSSION TITLE
Visualize spatial statistics with quadkeys

### DISCUSSION THREAD

**Discussion by wipfli** [onthegomap/planetiler#540] [Q&A]
Quadkeys or Bing maps tile system are useful to make spatial statistics, for example one can count how many buildings are in a given region.

I think we could assemble spatial statistics in the pre process step with quadkeys. And then we could render the tiles with the stats.

Is it possible to inject geometries after the pre-processed step?

**Comment by wipfli** [comment]
Other tiling systems like for example, h3 could also look great...

**Comment by bdon** [comment]
I think this is possible by holding in-memory state in your Profile layer implementations, and maybe not part of planetiler itself; the APIs in planetiler mostly concern objects that will need to be disk-backed because of storage size. If you are calculating statistics for H3 hex cells or quad keys, those seem pretty small (64 bit int -> 64 bit int).

when you call `postProcess` you can query that in-memory data struct to create new geometries; one issue however is that the current ForwardingProfile post-processing only exposes a `zoom` to the code and not a z/x/y. So I think that may be a useful enhancement?

**Reply by wipfli** [reply]
Nice. I would like to give that a try. Maybe postprocess can emit h3 polygons. For buildings, resolution 9 might be good. A h3 tile could be activated if say more than 5 buildings are in it...

![Screenshot_20230329-001423.png](https://user-images.githubusercontent.com/53421382/228379149-2439e5c9-2a68-40ee-9768-e595fea5f92d.png)

**Reply by msbarry** [reply]
Yeah I think the ideal solution is if you can make it so the output geometries you are aggregating by fit cleanly onto tiles (possibly with a larger tile buffer) - then planetiler automatically pulls all data on a tile together and you could do a smaller aggregation by geometry in memory in postProcess and emit new features there.

If they don't fit cleanly into tiles then you could store in memory on the profile and emit the new geometries in `finish` like the openmaptiles boundary layer does: https://github.com/openmaptiles/planetiler-openmaptiles/blob/5f2811895f11ce36a0ecefe0b2c975e7ca3071b6/src/main/java/org/openmaptiles/layers/Boundary.java#L309
